python/pyspark/ml/tests.py

```diff
@@ -674,8 +676,11 @@ def _compare_pipelines(self, m1, m2):
         if isinstance(m1, JavaParams):
             self.assertEqual(len(m1.params), len(m2.params))
             for p in m1.params:
-                self.assertEqual(m1.getOrDefault(p), m2.getOrDefault(p))
-                self.assertEqual(p.parent, m2.getParam(p.name).parent)
+                # Prevent key not found error in case of some param neither in paramMap and
+                # defaultParamMap.
+                if p in m1._paramMap or p in m1._defaultParamMap:
```

Comment:
Use ```m1.isDefined```.
Also, how about creating a helper method:
```
def _compare_param(m1, m2, paramName)
```
which implements these 3 lines?  That should eliminate some duplicate code.